File: Controls/ControlExif.ascx.vb
'
' DotNetNuke® - http://www.dotnetnuke.com
' Copyright (c) 2002-2011 by DotNetNuke Corp. 

'
' Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
' documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
' the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
' to permit persons to whom the Software is furnished to do so, subject to the following conditions:
'
' The above copyright notice and this permission notice shall be included in all copies or substantial portions 
' of the Software.
'
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
' TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
' THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
' CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
' DEALINGS IN THE SOFTWARE.
'
Option Strict On

Imports System
Imports System.Configuration
Imports DotNetNuke.Entities.Portals
Imports DotNetNuke.UI.UserControls
Imports DotNetNuke.Services.Localization
Imports Gallery.Exif

Namespace DotNetNuke.Modules.Gallery.WebControls

  Public Class Exif
    Inherits GalleryWebControlBase

#Region "Private Members"
    Private _CurrentRequest As GalleryViewerRequest = Nothing

    '/ <summary>An instance of the PhotoProperties class</summary>
    Private mPhotoProps As PhotoProperties
    '/ <summary>Has the PhotoProperties instance successfully initialized?</summary>
    Private mPhotoPropsInitialized As Boolean
    Private mIsAnalysisSuccessful As Boolean
    Private mPhotoPath As String = ""
    Private mCurrentItem As GalleryFile
#End Region

#Region "Public Properties"
    Public Property CurrentRequest() As GalleryViewerRequest
      Get
        Return _CurrentRequest
      End Get
      Set(ByVal value As GalleryViewerRequest)
        _CurrentRequest = value
      End Set
    End Property
#End Region

#Region "Private Methods"
    Protected Function ExifHelp(ByVal tagDatumName As String) As String
      Return Localization.GetString(tagDatumName & ".Help", Me.LocalResourceFile)
    End Function

    Protected Function ExifText(ByVal tagDatumName As String) As String
      Return Localization.GetString(tagDatumName & ".Text", Me.LocalResourceFile)
    End Function

    Private Sub BindList()
      Services.Localization.Localization.LocalizeDataGrid(grdExif, Me.LocalResourceFile)
      grdExif.DataSource = mPhotoProps.PhotoMetaData
      grdExif.DataBind()
    End Sub

    '/ <summary>
    '/ Initializes the PhotoProperties tag properties using asynchronous 
    '/ method invocation of the initialization.</summary>
    Private Sub InitializePhotoProperties(ByVal initXmlFile As String)
      ' Create an instance of the PhotoProperties
      mPhotoProps = New PhotoProperties
      mPhotoPropsInitialized = mPhotoProps.Initialize(initXmlFile)

    End Sub 'InitializePhotoProperties

    Private Function AnalyzeImageFile(ByVal fileName As String) As Boolean
      Dim isAnalyzed As Boolean = False
      Try
        mPhotoProps.Analyze(fileName)
        isAnalyzed = True
      Catch ex As InvalidOperationException
      End Try
      Return isAnalyzed
    End Function 'AnalyzeImageFile

    Private Sub InitExifData()
      Dim xmlFile As String = IO.Path.Combine(Server.MapPath(GalleryConfig.SourceDirectory & "/Resources"), "_photoMetadata.xml")
      InitializePhotoProperties(xmlFile)

      mIsAnalysisSuccessful = AnalyzeImageFile(mPhotoPath)
      If mIsAnalysisSuccessful Then
        BindList()
      End If

    End Sub
#End Region

#Region "Event Handlers"
    Private Sub Page_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
      CurrentRequest = New GalleryViewerRequest(ModuleId)
      mCurrentItem = CurrentRequest.CurrentItem
      mPhotoPath = mCurrentItem.SourcePath

      If Not IO.File.Exists(mPhotoPath) Then
        mPhotoPath = mCurrentItem.Path
      End If

      InitExifData()
      Dim isPopup As Boolean = Me.Parent.TemplateControl.AppRelativeVirtualPath.EndsWith("aspx")

      If isPopup Then
        Dim GalleryPage As GalleryPageBase = CType(Me.Parent.TemplateControl, GalleryPageBase)
        With GalleryPage
          .Title = GalleryPage.Title & " > " & CurrentRequest.CurrentItem.Title
          .ControlTitle = CurrentRequest.CurrentItem.Title
        End With
      Else
        Dim SitePage As CDefault = CType(Me.Page, CDefault)
        SitePage.Title = SitePage.Title & " > " & CurrentRequest.CurrentItem.Title
        Dim namingContainer As Gallery.ExifMetaData = CType(Me.NamingContainer, Gallery.ExifMetaData)
        With namingContainer
          .Title = CurrentRequest.CurrentItem.Title
        End With
      End If
      imgExif.ImageUrl = mCurrentItem.ThumbnailURL
      litInfo.Text = "<span>" & CurrentRequest.CurrentItem.ItemInfo & "</span>"
      litDescription.Text = "<p class=""Gallery_Description"">" & CurrentRequest.CurrentItem.Description & "</p>"
    End Sub

    Private Sub grdExif_ItemDataBound(sender As Object, e As System.Web.UI.WebControls.DataGridItemEventArgs) Handles grdExif.ItemDataBound
      If e.Item.ItemType = ListItemType.Item Or e.Item.ItemType = ListItemType.AlternatingItem Then
        Dim dataItem As PhotoTagDatum = DirectCast(e.Item.DataItem, PhotoTagDatum)
        If dataItem IsNot Nothing AndAlso dataItem.Id = 37500 Then
          'Remove MakerNote Tag (ID=37500) as it can cantain thousands of Hex values which are propriatary to the camera maker and often encrypted
          'Including the MakerNote Tag value in the grid causes a several minute delay before the grid is rendered with some photo images
          e.Item.Cells(3).Text = "&nbsp;"
        End If
      End If
    End Sub
#End Region

#Region " Web Form Designer Generated Code "

    'This call is required by the Web Form Designer.
    <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()

    End Sub


    'NOTE: The following placeholder declaration is required by the Web Form Designer.
    'Do not delete or move it.
    Private designerPlaceholderDeclaration As System.Object

    Private Sub Page_Init(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Init
      'CODEGEN: This method call is required by the Web Form Designer
      'Do not modify it using the code editor.
      InitializeComponent()
    End Sub

#End Region

  End Class

End Namespace

File: Controls/ControlExif.ascx.cs
using DotNetNuke;
using DotNetNuke.Common;
using DotNetNuke.Common.Utilities;
using DotNetNuke.Entities;
using DotNetNuke.Framework;
using DotNetNuke.Security;
using DotNetNuke.Services;
using DotNetNuke.Services.Exceptions;
using DotNetNuke.Services.Localization;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.Collections.Specialized;
using System.Configuration;
using System.Text;
using System.Text.RegularExpressions;
using System.Web;
using System.Web.Caching;
using System.Web.SessionState;
using System.Web.Security;
using System.Web.Profile;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;
//
// DotNetNuke® - http://www.dotnetnuke.com
// Copyright (c) 2002-2011 by DotNetNuke Corp. 

//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions 
// of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
// DEALINGS IN THE SOFTWARE.
//

using DotNetNuke.Entities.Portals;
using DotNetNuke.UI.UserControls;
using Gallery.Exif;

namespace DotNetNuke.Modules.Gallery.WebControls
{

	public partial class Exif : GalleryWebControlBase
	{

		#region "Private Members"

		private GalleryViewerRequest _CurrentRequest = null;
		/// <summary>An instance of the PhotoProperties class</summary>
		private PhotoProperties mPhotoProps;
		/// <summary>Has the PhotoProperties instance successfully initialized?</summary>
		private bool mPhotoPropsInitialized;
		private bool mIsAnalysisSuccessful;
		private string mPhotoPath = "";
			#endregion
		private GalleryFile mCurrentItem;

		#region "Public Properties"
		public GalleryViewerRequest CurrentRequest {
			get { return _CurrentRequest; }
			set { _CurrentRequest = value; }
		}
		#endregion

		#region "Private Methods"
		protected string ExifHelp(string tagDatumName)
		{
			return Localization.GetString(tagDatumName + ".Help", this.LocalResourceFile);
		}

		protected string ExifText(string tagDatumName)
		{
			return Localization.GetString(tagDatumName + ".Text", this.LocalResourceFile);
		}

		private void BindList()
		{
			DotNetNuke.Services.Localization.Localization.LocalizeDataGrid(ref grdExif, this.LocalResourceFile);
			grdExif.DataSource = mPhotoProps.PhotoMetaData;
			grdExif.DataBind();
		}

		/// <summary>
		/// Initializes the PhotoProperties tag properties using asynchronous 
		/// method invocation of the initialization.</summary>
		private void InitializePhotoProperties(string initXmlFile)
		{
			// Create an instance of the PhotoProperties
			mPhotoProps = new PhotoProperties();
			mPhotoPropsInitialized = mPhotoProps.Initialize(initXmlFile);

		}
		//InitializePhotoProperties

		private bool AnalyzeImageFile(string fileName)
		{
			bool isAnalyzed = false;
			try {
				mPhotoProps.Analyze(fileName);
				isAnalyzed = true;
			} catch (InvalidOperationException ex) {
			}
			return isAnalyzed;
		}
		//AnalyzeImageFile

		private void InitExifData()
		{
			string xmlFile = System.IO.Path.Combine(Server.MapPath(GalleryConfig.SourceDirectory() + "/Resources"), "_photoMetadata.xml");
			InitializePhotoProperties(xmlFile);

			mIsAnalysisSuccessful = AnalyzeImageFile(mPhotoPath);
			if (mIsAnalysisSuccessful) {
				BindList();
			}

		}
		#endregion

		#region "Event Handlers"
		private void Page_Load(System.Object sender, System.EventArgs e)
		{
			CurrentRequest = new GalleryViewerRequest(ModuleId);
			mCurrentItem = CurrentRequest.CurrentItem;
			mPhotoPath = mCurrentItem.SourcePath;

			if (!System.IO.File.Exists(mPhotoPath)) {
				mPhotoPath = mCurrentItem.Path;
			}

			InitExifData();
			bool isPopup = this.Parent.TemplateControl.AppRelativeVirtualPath.EndsWith("aspx");

			if (isPopup) {
				GalleryPageBase GalleryPage = (GalleryPageBase)this.Parent.TemplateControl;
				var _with1 = GalleryPage;
				_with1.Title = GalleryPage.Title + " > " + CurrentRequest.CurrentItem.Title;
				_with1.ControlTitle = CurrentRequest.CurrentItem.Title;
			} else {
				CDefault SitePage = (CDefault)this.Page;
				SitePage.Title = SitePage.Title + " > " + CurrentRequest.CurrentItem.Title;
				Gallery.ExifMetaData namingContainer = (Gallery.ExifMetaData)this.NamingContainer;
				var _with2 = namingContainer;
				_with2.Title = CurrentRequest.CurrentItem.Title;
			}
			imgExif.ImageUrl = mCurrentItem.ThumbnailURL;
			litInfo.Text = "<span>" + CurrentRequest.CurrentItem.ItemInfo + "</span>";
			litDescription.Text = "<p class=\"Gallery_Description\">" + CurrentRequest.CurrentItem.Description + "</p>";
		}

		private void grdExif_ItemDataBound(object sender, System.Web.UI.WebControls.DataGridItemEventArgs e)
		{
			if (e.Item.ItemType == ListItemType.Item | e.Item.ItemType == ListItemType.AlternatingItem) {
				PhotoTagDatum dataItem = (PhotoTagDatum)e.Item.DataItem;
				if (dataItem != null && dataItem.Id == 37500) {
					//Remove MakerNote Tag (ID=37500) as it can cantain thousands of Hex values which are propriatary to the camera maker and often encrypted
					//Including the MakerNote Tag value in the grid causes a several minute delay before the grid is rendered with some photo images
					e.Item.Cells[3].Text = "&nbsp;";
				}
			}
		}
		#endregion

		#region " Web Form Designer Generated Code "

		//This call is required by the Web Form Designer.
		[System.Diagnostics.DebuggerStepThrough()]

		private void InitializeComponent()
		{
		}


		//NOTE: The following placeholder declaration is required by the Web Form Designer.
		//Do not delete or move it.

		private System.Object designerPlaceholderDeclaration;
		private void Page_Init(System.Object sender, System.EventArgs e)
		{
			//CODEGEN: This method call is required by the Web Form Designer
			//Do not modify it using the code editor.
			InitializeComponent();
		}
		public Exif()
		{
			Init += Page_Init;
			Load += Page_Load;
		}

		#endregion

	}

}
